### DISCUSSION TITLE
How do I  suspend， continue or cancle task when use S3 multipartUpload

### DISCUSSION THREAD

**Discussion by yjjandyxy** [soto-project/soto#616] [Q&A] [answered]
How do I  suspend， continue or cancel task when using S3 multipart upload

**Comment by adam-fowler** [comment, marked as answer]
You can throw an error from the progress function to cancel a multipart upload. 

If you called `multipartUpload` with `abortOnFail` set to true it will abort the multipart upload and clean up.

If you call `multipartUpload` with `abortOnFail` set to `false` instead of aborting the multipart upload on an error being thrown it throws a new error type `S3ErrorType.multipart.abortedUpload` which includes all the information you need to resume the upload in a `S3.ResumeMultipartUploadRequest` and the error that caused the upload to fail. You can  then use the `S3.ResumeMultipartUploadRequest` object in `resumeMultipartUpload` to resume the multipart upload.

You can find out more [here](https://soto.codes/user-guides/s3-multipart-upload.html)

**Reply by yjjandyxy** [reply]
cancel task，I don't know much about it， Can you give me a demo? or give me some code

**Reply by adam-fowler** [reply]
You could do something similar to this. When you set `cancelMultipartUpload` to true it will cancel the multipart operation. Currently Soto doesn't support Swift concurrency cancellation using `Task.cancel`. It is on the list of things to fix.
```swift
import SotoS3

struct MultipartCancellationError: Error {}

var cancelMultipartUpload = false
let client = AWSClient(httpClientProvider: .createNew)
let s3 = S3(client: client)

let request = S3.CreateMultipartUploadRequest(bucket: "myBucket", key: "myKey")
let result = try s3.multipartUpload(request, filename: "tests.crash", abortOnFail: false) { progress in
    if cancelMultipartUpload {
        throw MultipartCancellationError()
    }
    print(progress)
}.wait()
```

**Reply by yjjandyxy** [reply]
Thanks
